packages/tools-iottwinmaker/src/lib/workspace.ts

```diff
@@ -318,5 +326,142 @@ async function createWorkspaceIfNotExists(workspaceId: string) {
+      console.log('Logging Bucket not found, moving on with deletion.');
+    } else {
+      throw e;
+    }
+  }
+  try {
+    console.log('Deleting workspace bucket...');
+    await deleteS3Bucket(s3BucketName);
+  } catch (e) {
+    if (e instanceof NoSuchBucket) {
+      console.log(`Bucket does not exist and cannot be deleted: ${s3BucketName}`);
+    } else {
+      throw e;
+    }
+  }
+}
+
+/**
+ * Deletes a role and any associated policies following the steps outlined here:
+ * https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_manage_delete.html#roles-managingrole-deleting-cli
+ * @param roleName role name
+ */
+async function deleteIAMroleAndPolicy(roleName: string | undefined): Promise<void> {
+  // Step 1: Remove the role from all instance profiles that the role is associated with
+  const instanceProfileResp = await aws().iam.listInstanceProfilesForRole({ RoleName: roleName });
+  if (instanceProfileResp.InstanceProfiles != undefined) {
+    for (const instanceProfile of instanceProfileResp.InstanceProfiles) {
+      await aws().iam.removeRoleFromInstanceProfile({
+        RoleName: roleName,
+        InstanceProfileName: instanceProfile.InstanceProfileName,
+      });
+      console.log(`removed role from instance profile: ${instanceProfile.InstanceProfileName}`);
+    }
+  }
+  // Step 2: Delete all policies that are associated with the role
+  const rolePolicies = await aws().iam.listRolePolicies({ RoleName: roleName });
+  if (rolePolicies.PolicyNames != undefined) {
+    for (const policy of rolePolicies.PolicyNames) {
+      await aws().iam.deleteRolePolicy({ RoleName: roleName, PolicyName: policy });
+      console.log(`deleted policy: ${policy}`);
+    }
+  }
+  const attachedRolePolicies = await aws().iam.listAttachedRolePolicies({ RoleName: roleName });
+  if (attachedRolePolicies.AttachedPolicies != undefined) {
+    for (const attachedPolicy of attachedRolePolicies.AttachedPolicies) {
+      await aws().iam.detachRolePolicy({ RoleName: roleName, PolicyArn: attachedPolicy.PolicyArn });
+      await aws().iam.deletePolicy({ PolicyArn: attachedPolicy.PolicyArn });
+      console.log(`detached and deleted policy: ${attachedPolicy.PolicyName}`);
+    }
+  }
+  // Step 3: Delete role
+  await aws().iam.deleteRole({ RoleName: roleName });
+  console.log(`deleted role: ${roleName}`);
+}
+
+/**
+ * Deletes a specified TwinMaker workspace
+ * @param workspaceId TM workspaceID
+ * @returns DeleteWorkspaceCommandOutput if successful
+ */
+async function deleteWorkspace(workspaceId: string): Promise<DeleteWorkspaceCommandOutput | undefined> {
+  try {
+    const deleteWorkspaceResp: DeleteWorkspaceCommandOutput = await aws().tm.deleteWorkspace({ workspaceId });
+    console.log(`deleted workspace: ${workspaceId}`);
+    return deleteWorkspaceResp;
+  } catch (err) {
+    console.log(`Unable to delete workspace ${workspaceId}:\n ${err}`);
```

Comment:
Should we just throw an error here? If customer stated to delete workspace and it fails we should make that obvious

Reply:
What was the discrepancy in verify? Looks like it catches to give custom message but does throw the error back out? https://github.com/awslabs/iot-app-kit/blob/main/packages/tools-iottwinmaker/src/lib/utils.ts#L19

If we just catch and log will the script return a success (e.g. `0`) error code (which doesn't seem right?)

Reply:
There is no discrepancy right now (sorry for confusion), we just had a similar change in that function (I was told in a JS course that it's always best practice to catch any known potential errors such as API calls). But now that you mention it, returning the 0 error code doesn't make sense, you're right I will update.